lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp

```diff
@@ -354,6 +355,37 @@ struct WarpOpDeadResult : public OpRewritePattern<WarpSingleLaneOp> {
   }
 };
 
+// If an operand is directly yielded out of the region we can forward it
+// directly and it doesn't need to go through the region.
+struct WarpOpForwardOperand : public OpRewritePattern<WarpSingleLaneOp> {
+  using OpRewritePattern<WarpSingleLaneOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(WarpSingleLaneOp warpOp,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<Type> resultTypes;
+    SmallVector<Value> yieldValues;
+    auto yield = cast<vector_ext::YieldOp>(
+        warpOp.getBodyRegion().getBlocks().begin()->getTerminator());
+    Value valForwarded;
+    unsigned resultIndex;
+    for (OpOperand &operand : yield->getOpOperands()) {
+      auto arg = operand.get().dyn_cast<BlockArgument>();
+      if (!arg || arg.getOwner()->getParentOp() != warpOp.getOperation())
+        continue;
+      Value result = warpOp.getResult(operand.getOperandNumber());
+      Value warpOperand = warpOp.args()[arg.getArgNumber()];
+      if (result.use_empty() || result.getType() != warpOperand.getType())
+        continue;
+      valForwarded = warpOperand;
+      resultIndex = operand.getOperandNumber();
+      break;
+    }
+    if (!valForwarded)
+      return failure();
+    warpOp.getResult(resultIndex).replaceAllUsesWith(valForwarded);
```

Comment:
Is this safe? I always thought we would have to go through the rewriter API or do something like `updateRootInplace` (which does not work with `replaceAllUsesWith`).

Reply:
This doesn't modify the root so I though we didn't need `updateRootInplace` here. I see some other examples upstream which seem to do something similar so I assumed it was correct:
https://github.com/llvm/llvm-project/blob/5d47332783d0421ba0da2a723354fb523eb85d0b/mlir/lib/Dialect/SCF/SCF.cpp#L1449

Although I'm not completely clear if this case is allowed